clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java

```diff
@@ -210,6 +220,38 @@ public void testMultipleServerMechanisms() throws Exception {
     }
 
     /**
+     * Tests that Kafka ApiVersionRequests are handled by the SASL server authenticator
+     * prior to SASL handshake flow and that subsequent authentication succeeds.
+     */
+    @Test
+    public void testUnauthenticatedApiVersionRequest() throws Exception {
+        SecurityProtocol securityProtocol = SecurityProtocol.SASL_PLAINTEXT;
+        configureMechanisms("PLAIN", Arrays.asList("PLAIN"));
+        server = NetworkTestUtils.createEchoServer(securityProtocol, saslServerConfigs);
+
+        // Create PLAINTEXT connection to manually authenticate after ApiVersionsRequest
+        String node = "1";
+        createClientConnection(SecurityProtocol.PLAINTEXT, node);
```

Comment:
@gwenshap I will see if I can add an integration test as well tomorrow (it is nearly 1am here, so I will do it tomorrow morning). But just to be clear, the client in the current unit test does not use SASL. It uses PLAINTEXT. It does not send a handshake request before ApiVersionsRequest, and if it did, the test would fail because it relies entirely on the Sasl authenticator to respond to the handshake and version requests. There is a second unit test which verifies that too.